Fix FlowableOnBackpressureBufferStrategy

matgabriel opened the pull request:
Fix buffered objects not propagated downstream in FlowableOnBackpressureBufferStrategy
Fix drop strategy logic in FlowableOnBackpressureBufferStrategy
Add unit test for FlowableOnBackpressureBufferStrategy, copied from FlowableOnBackpressureBufferTest, there is still some work needed to have a better coverage

matgabriel committed:
Fix buffered objects not propagated downstream in FlowableOnBackpressureBufferStrategy
Fix drop strategy logic in FlowableOnBackpressureBufferStrategy
Add unit test for FlowableOnBackpressureBufferStrategy, copied from FlowableOnBackpressureBufferTest, there is still some work needed to have a better coverage

akarnokd commented on src/test/java/io/reactivex/internal/operators/flowable/FlowableOnBackpressureBufferStrategyTest.java:
```diff
@@ -0,0 +1,133 @@
+
+public class FlowableOnBackpressureBufferStrategyTest {
+
+    private static Action onOverFlow = new Action() {
```
Use `Functions.EMPTY_ACTION`.

akarnokd commented on src/test/java/io/reactivex/internal/operators/flowable/FlowableOnBackpressureBufferStrategyTest.java:
```diff
@@ -0,0 +1,133 @@
+
+
+    @Test(timeout = 2000)
+    public void testFixBackpressureWithBuffer() throws InterruptedException {
```
No need to go async, just don't request from the TestSubscriber.

matgabriel replied:
What do you mean exactly?
If we don't subscribe on another thread and we don't request from the TestSubscriber, then the main thread will be stuck on the 'subscribe(ts)'

akarnokd replied:
Don't request upfront. Fill the buffer until it drops items then issue a `TestSubscriber.request()` and see if the right value comes out. No threading needed.

akarnokd commented on src/test/java/io/reactivex/internal/operators/flowable/FlowableOnBackpressureBufferStrategyTest.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testFixBackpressureBufferNegativeCapacity() throws InterruptedException {
```
Please drop `test`, `testFix` and `fix` prefixes.

akarnokd commented on src/main/java/io/reactivex/internal/operators/flowable/FlowableOnBackpressureBufferStrategy.java:
```diff
@@ -123,6 +123,7 @@ public void onNext(T t) {
                    }
                } else {
                    dq.offer(t);
+                   drain();
```
Should not call drain while in the synchronized block. It should look like this:

``` java
            if (done) {
                return;
            }
            boolean callOnOverflow = false;
            boolean callError = false;
            Deque<T> dq = deque;
            synchronized (dq) {
               if (dq.size() == bufferSize) {
                   switch (strategy) {
                   case DROP_LATEST:
                       dq.pollLast();
                       dq.offer(t);
                       callOnOverflow = true;
                       break;
                   case DROP_OLDEST:
                       dq.poll();
                       dq.offer(t);
                       callOnOverflow = true;
                       break;
                   default:
                       // signal error
                       callError = true;
                       break;
                   }
               } else {
                   dq.offer(t);
                   return;
               }
            }

            if (callOnOverflow) {
                if (onOverflow != null) {
                    try {
                        onOverflow.run();
                    } catch (Throwable ex) {
                        Exceptions.throwIfFatal(ex);
                        s.cancel();
                        onError(ex);
                    }
                }
            } 
            else if (callError) {
                s.cancel();
                onError(new MissingBackpressureException());
            } else {
                drain();
            }
```

matgabriel committed:
Fix FlowableOnBackpressureBufferStrategy

Move the call to drain out of the synchronized block
Cleaned up unit tests to follow recommendations from akarnokd

akarnokd commented:
Thanks! :+1:
